Instantiate the Jenkins template when JenkinsPipelineStrategy is created

mfojtik opened the pull request.

mfojtik commented:
@jimmidyson don't merge this pls :-)

mfojtik commented:
@bparees @smarterclayton PTAL

bparees commented on pkg/build/client/clients.go:
```diff
@@ -96,3 +101,18 @@ func NewOSClientBuildConfigInstantiatorClient(client osclient.Interface) *OSClie
+	return &OSClientPipelineBuildConfigInstantiatorClient{Client: client, KubeClient: kubeClient}
+}
+
+// Instantiate generates new build for given buildConfig
```
fix godoc

bparees commented on pkg/build/client/clients.go:
```diff
@@ -96,3 +101,18 @@ func NewOSClientBuildConfigInstantiatorClient(client osclient.Interface) *OSClie
+	KubeClient kclient.Interface
+}
+
+// NewOSClientBuildConfigInstantiatorClient creates a new build client that uses an openshift client to create builds
```
fix godoc

bparees commented on pkg/build/controller/factory/factory.go:
```diff
@@ -335,6 +335,39 @@ func (factory *BuildConfigControllerFactory) Create() controller.RunnableControl
 	}
 }
 
+type PipelineControllerFactory struct {
```
why can't we reuse the existing buildconfig controller and just add the logic for creating jenkins there?

jimmidyson commented:
@mfojtik I've rebased the branch against Origin master so you'll need to rebase too to have a sane PR...

@bparees @mfojtik Feel free to merge when you're happy - you should have karma to do so.

mfojtik commented:
@jimmidyson thanks! will do some rounds of reviews from Ben and Clayton first before merging. I already rebased :-)

bparees commented on pkg/build/controller/pipeline_controller.go:
```diff
@@ -0,0 +1,60 @@
+
+// JenkinsPipelineServiceName is the name of the Jenkins service this pipeline
+// deploys.
+const JenkinsPipelineServiceName = "ci"
```
since we're being specific, we should include jenkins in this name.  jenkins-ci (if that's a legal name)?

bparees commented on pkg/build/controller/pipeline_controller.go:
```diff
@@ -0,0 +1,60 @@
+// If the project where this configuration is create does not already have the
+// "jenkins" service, this will deploy a template which provision a new Jenkins
+// instance.
+func (c *PipelineController) HandlePipeline(bc *buildapi.BuildConfig) error {
```
assuming you keep this, it's really HandleBuildConfig still, since that is what it is handling.

bparees commented on pkg/build/controller/pipeline_controller.go:
```diff
@@ -0,0 +1,60 @@
+	if bc.Status.LastVersion > 0 {
+		glog.V(4).Infof("LastVersion is %d", bc.Status.LastVersion)
+		return nil
+	}
```
what's the logic behind aborting here?  it shouldn't matter what the BC last version is, if there's no jenkins service, we need to create one, right?

smarterclayton replied:
Yes, we do.

bparees commented on pkg/build/controller/pipeline_controller.go:
```diff
@@ -0,0 +1,60 @@
+
+	// TODO: Make the service name configurable
+	if _, err := c.KubeClient.Services(bc.Namespace).Get(JenkinsPipelineServiceName); err == nil {
+		glog.V(1).Infof("The jenkins service already exists in project %q", bc.Namespace)
```
feels like a V(4) to me.  this is going to show up a lot (every time a re-list happens)

bparees commented on pkg/build/controller/pipeline_controller.go:
```diff
@@ -0,0 +1,60 @@
+		return nil
+	}
+
+	glog.V(4).Infof("Provisioning new jenkins service for project %q", bc.Namespace)
```
whereas this seems like a V(1) or V(2)

smarterclayton replied:
v(3), even normal controller function isn't newsworthy.

bparees commented on pkg/client/buildconfigs.go:
```diff
@@ -31,6 +42,7 @@ type BuildConfigInterface interface {
 
 	Instantiate(request *buildapi.BuildRequest) (result *buildapi.Build, err error)
 	InstantiateBinary(request *buildapi.BinaryBuildRequestOptions, r io.Reader) (result *buildapi.Build, err error)
+	InstantiatePipeline(c *buildapi.JenkinsPipelineBuildStrategy, k kclient.Interface) error
```
i'm not convinced this should be hung off the BuildConfig.  Creating the jenkins service is more a side-effect of the build config being created, than something being created from the buildconfig (which is what instantiate and instantiatebinary are doing)

smarterclayton replied:
This doesn't belong here.

bparees commented on pkg/client/buildconfigs.go:
```diff
@@ -130,3 +142,84 @@ func (c *buildConfigs) InstantiateBinary(request *buildapi.BinaryBuildRequestOpt
+}
+
+func (c *buildConfigs) InstantiatePipeline(config *buildapi.JenkinsPipelineBuildStrategy, k kclient.Interface) error {
+	// TODO: Should the template be configured at strategy level?
```
no.  possibly at the cluster level or project level, but definitely not at the buildconfig strategy level.

smarterclayton replied:
This is some sort of helper library that is associated with the controller and the master config, not with clients.

mfojtik commented:
@bparees @smarterclayton refactored a little bit :-)
- Removed pipeline_controller (still like the idea of having separate controller for instantiating Jenkins, but I guess this solution will save us from having yet another etcd watcher/cache/etc.)
- Moved all Jenkins template instantiation into separate helper (pkg/build/util/jenkins.go)
- Fixed glog levels and a lot of small nits/godocs.

Right now I have the template namespace/name and service name as golang consts. I agree with @smarterclayton that they should be part of the master config, where admin can set the default Jenkins template name and perhaps the service name as well. The service name must match with the one in the template, which I found a little bit funky (if they does not match, every time you change the BC, the processing/deploying happens again... which is no-op if you don't add more items into template).

jimmidyson commented on pkg/build/api/types.go:
```diff
@@ -30,6 +30,9 @@ const (
 	DropCapabilities = "DROP_CAPS"
+	// JenkinsDefaultServiceName is the default name of the Jenkins service deployed by
+	// JenkinsPipelineBuildStrategy.
+	JenkinsDefaultServiceName = "ci"
```
Wouldn't `jenkins` be better here?

jstrachan replied:
I'd have thought 'jenkins' was a better default service name? Folks may already be using Bamboo / Teamcity / something-else as their CI server?

mfojtik replied:
@jstrachan `ci` was @smarterclayton suggestion... i agree with you that `jenkins-ci` would be a better name. Users might already have 'jenkins' server running, so maybe being explicit with `-ci` suffix isn't bad idea.

jimmidyson replied:
openshift-jenkins? managed-jenkins? Something that differentiates from a potential user Jenkins would be good, although impossible to enforce.

jimmidyson commented on pkg/build/controller/config_controller.go:
```diff
@@ -33,11 +37,32 @@ func IsFatal(err error) bool {
 
+	if strategy := bc.Spec.Strategy.JenkinsPipelineStrategy; strategy != nil {
+		glog.V(4).Infof("Detected Jenkins pipeline strategy in %s/%s build configuration", bc.Namespace, bc.Name)
+		if _, err := c.KubeClient.Services(bc.Namespace).Get(buildapi.JenkinsDefaultServiceName); err == nil {
```
What if the error returned is e.g. unable to communicate with API server? How are retries handled?

mfojtik replied:
@jimmidyson i think the controller will handle retry, but I'm not 100% sure. @bparees any idea how the ConfigChange controller handle retries when something bad happens?

mfojtik replied:
@jimmidyson what I observed is that when I forgot to create the jenkins template, but I created this BC, nothing really happened. But as soon as I created the template it deployed automatically... I guess this will be the same case.

jimmidyson replied:
Do we want to check the err here to see if it is a bad error so this handler shouldn't continue & rely on whatever retry process is in place?

mfojtik replied:
@jimmidyson would be hard to decide what is "bad error" here... if the API cannot be reached, that might be temporary and should not prevent the controller from retrying.... same with non-existing template or broken template (might be fixed and redeployed)....

jimmidyson replied:
Can we check the status reason of the response to see if it's a not found & continue, log & return otherwise?

mfojtik replied:
@jimmidyson continue will mean that when response is not not found but an 503, the Jenkins will never be created unless you modify the build config and trigger another instantiation... I think we should keep retrying in case of any error. @bparees ?

mfojtik replied:
btw. I think we should at least report this as an event to user....

bparees replied:
the config controller is wrapped by a retry controller (like most of our controllers):
https://github.com/openshift/origin/blob/master/pkg/build/controller/factory/factory.go#L315-L335

so it all depends what HandleBuildConfig returns as an error, as to whether the event gets retried or not.  If you return a ConfigControllerFatalError, it won't.  otherwise it will.

it's hard to imagine a situation where we should not retry this since we should always expect to be able to reach the api server.  Note that the retry is limited, so eventually it does give up, but presumably the reget/list flow should start the retry again when that happens.

jimmidyson replied:
So I feel this should return an error & allow retry controller to do its thing if the API server is uncontactable.

mfojtik commented:
@bparees @jimmidyson added events for errors during instantiation/creation/etc.... also added event to tell users that jenkins was added without any errors...

mfojtik commented:
@jimmidyson @smarterclayton the Jenkins template name/namespace and svc name are now fully configurable in OpenShift master configuration. Lemme know if you are OK with defaults.

bparees commented on pkg/cmd/server/api/v1/conversions.go:
```diff
@@ -17,6 +17,15 @@ func addDefaultingFuncs(scheme *runtime.Scheme) {
+				obj.JenkinsPipelineConfig.TemplateName = "jenkins"
+			}
+			if len(obj.JenkinsPipelineConfig.ServiceName) == 0 {
+				obj.JenkinsPipelineConfig.ServiceName = "jenkins-ci"
```
dan made the point we should call this jenkins-pipeline or something more specific.

bparees replied:
probably the same for the template name

mfojtik replied:
fixed.

jstrachan replied:
jenkins 2.x comes with pipeline baked in so 'jenkins' is probably just fine for the service name?

bparees replied:
users may stand up their own jenkins service for other purposes, we don't want to steal that name from their namespace.

jstrachan replied:
I'd say its more likely that users in a project use 1 jenkins for pipelines and other builds than it is to run 2 jenkins, one for regular builds and one for pipelines. 

If folks want to run another jenkins they can, with another name?

bparees replied:
i don't think our infrastructure components should be consuming commonly desirable names from a user's namespace.  In an ideal world, the user might not even be aware this jenkins server/service exists.

smarterclayton replied:
`jenkins-pipeline` or `jenkins-ci` sound ok.  Remember that this is going to be part of the route name, so what route do we want on the outside for this?

`ci-namespace.myapps.com`?
`jenkins-ci-namespace.myapps.com`?

I'd probably prefer to go shorter rather than longer - we've been discussing having the git server take `git` in the namespace.

rawlingsj replied:
I think if we can avoid `-ci` in the name it might save us some pain later.  I seem to remember the Jenkins community trying to drop CI from their branding.

smarterclayton replied:
Another point here is that there is 1 of these per project, and the name is probably less jenkins specific than about the CI/CD workflow that we are creating.  So calling the service just "ci" or similar is more accurate (it's not about Jenkins necessarily, it's about the functionality it exposes in that project).

bparees replied:
let's remember that there are two scenarios we are targeting:

1) jenkinsfile pipelines executed by jenkins
2) "some pipeline definition" executed by "some pipeline engine"

this PR is targeted at (1) so it's appropriate to include jenkins in the names of the things it creates/relies on.

smarterclayton replied:
This PR is implementing both 1 and 2

On Wed, Mar 30, 2016 at 4:27 PM, Ben Parees notifications@github.com
wrote:

> In pkg/cmd/server/api/v1/conversions.go
> https://github.com/jimmidyson/origin/pull/2#discussion_r57957777:
> 
> > @@ -17,6 +17,15 @@ func addDefaultingFuncs(scheme *runtime.Scheme) {
> >             if len(obj.APILevels) == 0 {
> >                 obj.APILevels = internal.DefaultOpenShiftAPILevels
> >             }
> > -           if len(obj.JenkinsPipelineConfig.Namespace) == 0 {
> > -               obj.JenkinsPipelineConfig.Namespace = "openshift"
> > -           }
> > -           if len(obj.JenkinsPipelineConfig.TemplateName) == 0 {
> > -               obj.JenkinsPipelineConfig.TemplateName = "jenkins"
> > -           }
> > -           if len(obj.JenkinsPipelineConfig.ServiceName) == 0 {
> > -               obj.JenkinsPipelineConfig.ServiceName = "jenkins-ci"
> 
> let's remember that there are two scenarios we are targeting:
> 
> 1) jenkinsfile pipelines executed by jenkins
> 2) "some pipeline definition" executed by "some pipeline engine"
> 
> this PR is targeted at (1) so it's appropriate to include jenkins in the
> names of the things it creates/relies on.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/jimmidyson/origin/pull/2/files/3ed08d495dd391ab47317d2946c2695da23b317b#r57957777

smarterclayton replied:
The instance created by this PR is supposed to support both future
builders.  It is the generic "ci" engine for end users.  We will not create
two separate jenkins instances to support the two different types.  The
generic builder is a proxy to "I expect this build to be applied"

On Wed, Mar 30, 2016 at 4:28 PM, Clayton Coleman ccoleman@redhat.com
wrote:

> This PR is implementing both 1 and 2
> 
> On Wed, Mar 30, 2016 at 4:27 PM, Ben Parees notifications@github.com
> wrote:
> 
> > In pkg/cmd/server/api/v1/conversions.go
> > https://github.com/jimmidyson/origin/pull/2#discussion_r57957777:
> > 
> > > @@ -17,6 +17,15 @@ func addDefaultingFuncs(scheme *runtime.Scheme) {
> > >            if len(obj.APILevels) == 0 {
> > >                obj.APILevels = internal.DefaultOpenShiftAPILevels
> > >            }
> > > -          if len(obj.JenkinsPipelineConfig.Namespace) == 0 {
> > > -              obj.JenkinsPipelineConfig.Namespace = "openshift"
> > > -          }
> > > -          if len(obj.JenkinsPipelineConfig.TemplateName) == 0 {
> > > -              obj.JenkinsPipelineConfig.TemplateName = "jenkins"
> > > -          }
> > > -          if len(obj.JenkinsPipelineConfig.ServiceName) == 0 {
> > > -              obj.JenkinsPipelineConfig.ServiceName = "jenkins-ci"
> > 
> > let's remember that there are two scenarios we are targeting:
> > 
> > 1) jenkinsfile pipelines executed by jenkins
> > 2) "some pipeline definition" executed by "some pipeline engine"
> > 
> > this PR is targeted at (1) so it's appropriate to include jenkins in the
> > names of the things it creates/relies on.
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly or view it on GitHub
> > https://github.com/jimmidyson/origin/pull/2/files/3ed08d495dd391ab47317d2946c2695da23b317b#r57957777

bparees replied:
what happens when the underlying implementation engine for the other pipeline strategy is not jenkins?

smarterclayton replied:
That's why this name should probably be "ci".  We're installing the "ci"
component to your project (jenkins is a detail, a good one, but still a
detail)

On Wed, Mar 30, 2016 at 4:53 PM, Ben Parees notifications@github.com
wrote:

> In pkg/cmd/server/api/v1/conversions.go
> https://github.com/jimmidyson/origin/pull/2#discussion_r57961741:
> 
> > @@ -17,6 +17,15 @@ func addDefaultingFuncs(scheme *runtime.Scheme) {
> >             if len(obj.APILevels) == 0 {
> >                 obj.APILevels = internal.DefaultOpenShiftAPILevels
> >             }
> > -           if len(obj.JenkinsPipelineConfig.Namespace) == 0 {
> > -               obj.JenkinsPipelineConfig.Namespace = "openshift"
> > -           }
> > -           if len(obj.JenkinsPipelineConfig.TemplateName) == 0 {
> > -               obj.JenkinsPipelineConfig.TemplateName = "jenkins"
> > -           }
> > -           if len(obj.JenkinsPipelineConfig.ServiceName) == 0 {
> > -               obj.JenkinsPipelineConfig.ServiceName = "jenkins-ci"
> 
> what happens when the underlying implementation engine for the other
> pipeline strategy is not jenkins?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/jimmidyson/origin/pull/2/files/3ed08d495dd391ab47317d2946c2695da23b317b#r57961741

bparees replied:
what happens when i have both in a single project?

smarterclayton replied:
You can't have two auto CI installations in the same project.  If you
aren't using Jenkins to deliver pipelines, you aren't using Jenkins.

On Wed, Mar 30, 2016 at 5:32 PM, Ben Parees notifications@github.com
wrote:

> In pkg/cmd/server/api/v1/conversions.go
> https://github.com/jimmidyson/origin/pull/2#discussion_r57967728:
> 
> > @@ -17,6 +17,15 @@ func addDefaultingFuncs(scheme *runtime.Scheme) {
> >             if len(obj.APILevels) == 0 {
> >                 obj.APILevels = internal.DefaultOpenShiftAPILevels
> >             }
> > -           if len(obj.JenkinsPipelineConfig.Namespace) == 0 {
> > -               obj.JenkinsPipelineConfig.Namespace = "openshift"
> > -           }
> > -           if len(obj.JenkinsPipelineConfig.TemplateName) == 0 {
> > -               obj.JenkinsPipelineConfig.TemplateName = "jenkins"
> > -           }
> > -           if len(obj.JenkinsPipelineConfig.ServiceName) == 0 {
> > -               obj.JenkinsPipelineConfig.ServiceName = "jenkins-ci"
> 
> what happens when i have both in a single project?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/jimmidyson/origin/pull/2/files/3ed08d495dd391ab47317d2946c2695da23b317b#r57967728

smarterclayton replied:
The name of the variable in the configuration would be `DefaultContinuousIntegration` or something.  If the default CI doesn't support Jenkinsfile builders (in the future) we'd either add a flag to disable it, or simply make that a responsibility of the CI controller.

smarterclayton replied:
So we had a brief discussion afterwards.  Let's keep this JenkinsPipeline and `jenkins` for the short term future, we can always change it/refactor further on.  We'll label this an experimental feature anyway.

mfojtik replied:
@smarterclayton @bparees thanks! i changed the "jenkins-pipeline" to "jenkins".

mfojtik commented:
@bparees @smarterclayton any final comments before merging this to @jimmidyson PR?

mfojtik commented:
Merging this after IRC talk with Ben.
